stake/src/rate.rs

```diff
@@ -19,6 +26,51 @@ pub struct RateData {
     pub validator_exchange_rate: u64,
 }
 
+impl RateData {
+    pub fn next_rates(
+        &self,
+        base_rate_data: &BaseRateData,
+        funding_streams: Vec<FundingStream>,
+    ) -> RateData {
+        // compute the validator's total commissio
+        let commission_rate_bps = funding_streams
+            .iter()
+            .fold(0u64, |total, stream| total + stream.rate_bps as u64);
+
+        if commission_rate_bps > 1_0000 {
+            // we should never hit this branch: validator funding streams should be verified not to
+            // sum past 100% in the state machine's validation of registration of new funding
+            // streams
+            panic!("commission rate sums to > 100%")
+        }
+
+        // compute next validator reward rate
+        // 1 bps = 1e-4, so here we group digits by 4s rather than 3s as is usual
+        let validator_reward_rate =
+            ((1_0000_0000u64 - (commission_rate_bps * 1_0000)) * BASE_REWARD_RATE) / 1_0000_0000;
+
+        // compute validator exchange rate
+        let validator_exchange_rate = (self.validator_exchange_rate
+            * (self.validator_reward_rate + 1_0000_0000))
+            / 1_0000_0000;
+
+        // this is supposed to be multiplied by the number of delegation tokens,
+        // how do we track that?
+        //
+        // todo: consider specifying the voting power function as a pure function of current epoch
+        // state (delegation tokens, etc) instead of an adjustmenet function
+        let voting_power_adjustment =
+            (validator_exchange_rate * 1_0000_0000) / base_rate_data.base_exchange_rate;
+
+        RateData {
+            identity_key: self.identity_key.clone(),
+            epoch_index: self.epoch_index + 1,
+            voting_power: self.voting_power * voting_power_adjustment,
```

Comment:
@avahowell ah, wait, this is not correct, the voting power adjustment should be multiplied by the supply of delegation tokens, not by the current voting power (sorry for missing this in review!).

The previous code left the voting power untouched since we didn't have token supply tracking (#335), which is also not the intended behavior, but I think that this line would cause the voting power to blow up (if we ever committed it back to tendermint).

Rather than fixing this in place, we might as well just fix it when doing the calculations properly in #337.